### ISSUE TITLE
When cls.var: Type[T], the return type of cls.var() is inferred as Any

### ISSUE THREAD

**Issue by brettcs** [python/mypy#9361]
Minimal reproduction:

```python
from typing import (
    ClassVar,
    Generic,
    Type,
    TypeVar,
)

T = TypeVar('T')

class IFactory(Generic[T]):
    def __init__(self, t: Type[T]) -> None:
        self.t = t

    def build(self) -> T:
        return self.t()


class CFactory(Generic[T]):
    t: ClassVar[Type[T]]

    @classmethod
    def build(cls) -> T:
        return cls.t()
```

When running with Python 3.7.3 and mypy 0.770, `mypy --strict` reports:

```
classvar.py:23: error: Returning Any from function declared to return "T"  [no-any-return]
```

I expected mypy to be able to infer that the return type of `cls.t()` is `T`, since it does that just fine through an instance variable, as in IFactory here.

I get the same error whether or not I annotate the class variable with `ClassVar`.

**Comment by brettcs** [comment]
I just checked with the latest published mypy 0.782, and it reports the same error with the same code.

Can I please ask why this report was closed? Of course it's no problem if it's a duplicate or something, but if it's WONTFIX it would be helpful for my development to know that. Thanks in advance.

**Comment by gvanrossum** [comment]
The bug is in your code. Ask on Gitter for more help.
